src/Interpreters/Access/InterpreterExecuteAsQuery.cpp

```diff
@@ -0,0 +1,74 @@
+#include <Core/ServerSettings.h>
+#include <Interpreters/InterpreterFactory.h>
+#include <Interpreters/Access/InterpreterExecuteAsQuery.h>
+#include <Parsers/Access/ASTExecuteAsQuery.h>
+#include <Parsers/Access/ASTRolesOrUsersSet.h>
+#include <Access/RolesOrUsersSet.h>
+#include <Access/AccessControl.h>
+#include <Access/User.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterSelectWithUnionQuery.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int SUPPORT_IS_DISABLED;
+}
+
+namespace ServerSetting
+{
+    extern const ServerSettingsBool allow_impersonate_user;
+}
+
+
+BlockIO InterpreterExecuteAsQuery::execute()
+{
+    const auto current_user_uuid = getContext()->getSessionContext()->getUserID();
+    const auto & query = query_ptr->as<const ASTExecuteAsQuery &>();
+
+    setImpersonateUser(query);
+
+    BlockIO return_value = {};
+
+    auto restore_previous_user = [&] () { getContext()->getSessionContext()->switchImpersonateUser(RolesOrUsersSet(current_user_uuid.value())); };
```

Comment:
Here `restore_previous_user` won't work good enough - because `switchImpersonateUser()` also resets settings and roles. What is better is to clone the query context and you don't need to worry about restoring it afterwards:

```
if (query.subquery)
{
    auto impersonation_context = Context::createCopy(*getContext());
    impersonation_context->makeQueryContext();
    impersonation_context->switchImpersonateUser(...);
    return executeQuery(serializeAST(query.subquery), impersonation_context);
}
else
{
    getContext()->getSessionContext()->switchImpersonateUser();
    return {};
}